File: shell.lua
function Shell(id, x, y, velMag, angle)
  local s = {}
  s.id = id
  s.pos = Vec2(x, y)
  s.w = 2
  s.h = 4
  s.angle = angle
  s.velMag = velMag
  s.vel = Vec2(velMag*math.sin(angle), velMag*math.cos(angle))

  function s:draw()
    lg.push()
      lg.translate(self.pos.x, self.pos.y)
      lg.rotate(-self.angle)

      lg.setColor(1, 0, 0, 1)
      lg.rectangle("fill", -(self.w/2), -(self.h/2), self.w, self.h)
    lg.pop()
  end

  function s:update(dt)
    self.pos.x = self.pos.x + self.vel.x*dt
    self.pos.y = self.pos.y + self.vel.y*dt
  end

  table.insert(entities.proj, s)
  return s
end

File: src/entities/shell.go
package entities

import (
  "github.com/gen2brain/raylib-go/raylib"
	"tools"
)

const (
	SHELL_TIMEOUT float32 = 6 // Time in seconds
)

var (
	shellDrawFrame rl.Rectangle = rl.NewRectangle(0, 0, 6, 10)
	shellTex rl.Texture2D
)

type Shell struct {
	Projec
	Damage int
	Size rl.Vector2    // Where X is Width, Y is Height
}

func NewShell(IDnum int, pos, parentVel rl.Vector2, vel, angle float32, damage, W, H int) *Shell {
	if shellTex.ID == 0 {
		println("Loading shell texture.")
		shellTex = rl.LoadTexture("src/assets/shell/shell.png")
	}
	s := new(Shell)
  s.BaseBody = NewBody(NewID(IDnum, "shell"), pos, vel, angle)
	s.Damage = damage
	s.Size = rl.NewVector2(float32(W), float32(H))
	s.timeOfCreation = rl.GetTime()
	s.timeLimit = SHELL_TIMEOUT

	v := tools.GetXYComponent(s.VelMag, s.Angle)
	v.X += parentVel.X
	v.Y += parentVel.Y
	s.VelMag = tools.GetMagnitude(&v)
	return s
}

func (s *Shell) Draw() {
	rl.DrawTexturePro(shellTex, shellDrawFrame, rl.NewRectangle(s.Pos.X, s.Pos.Y, shellDrawFrame.Width, shellDrawFrame.Height), rl.NewVector2(shellDrawFrame.Width/2, shellDrawFrame.Height/2), s.Angle*rl.Rad2deg, rl.White)
}

func (s *Shell) Update(dt float32) {
	vel := tools.GetXYComponent(s.VelMag, s.Angle)
	s.Pos.X += (vel.X*dt)
	s.Pos.Y -= (vel.Y*dt)
}

